### ISSUE TITLE
Accuracy reduces, after converting Pytorch model to ONNX model

### ISSUE THREAD

**Issue by fsi-tk** [pytorch/pytorch#34204]
## Summary
I converted OSNet (https://github.com/KaiyangZhou/deep-person-reid) Pytorch model (.pth) to ONNX model (.onnx), using "torch.onnx.export". Then, I compared the outputs of Pytorch model and ONNX model. The errors between those outputs are 1e-1 in maximum. Because of these errors, Accuracy of ONNX model severely reduces (mAP less than 1%). I want to solve this problem to get as high accuracy as Pytorch model (mAP more than 80%).

#### 0. Error
The model I converted  is called as 'osnet_x1_0' in git repository I refered above. This model is used for person re-identification. This model converts an input image to a vector which has 512 elements as output.  The element-wise errors between both models are 1e-1 in maximum.

A part of outputs of both model is below. I think that each output shows same trend, but shows also large errors which cannot be ignored. I think acceptable error is 1e-5 at most.
```
Pytorch  ONNX    Error
-----------------------------------------
0.5647   0.6391   0.07430197
1.2318   1.2531   0.0212744
0.0000   0.0000   0   
1.8207   1.7937   -0.0269936
0.0000   0.0000   0   
0.2561   0.3273   0.07112018                                              
0.0000   0.0000   0   
0.3725   0.3347   -0.03780318
0.0000   0.0000   0   
0.1875   0.2940   0.10639236
0.0079   0.0000   -0.0079128
...
```

#### 1. Pre-processing
To reduce these errors, I confirm first that same pre-processing (showd below) are executed for input images of both model. I think nothing is wrong with pre-processing.

 - 255.0 division
 - Normalization using mean and std of ImageNet
 - Transpose to N, C, H, W order
 - Convert from BGR to RGB

#### 2. Implementation difference
Next, I tried getting rid of batch normalization layer from original OSNet Pytorch model, and converted this model to ONNX (because I had seen the case that the different implementation of instance normalization between two frameworks caused large error). Then, the outputs of these two models show around 1e-3 error.

This result may indicate that there are any differences between both frameworks implementations of batch normalization layer, and these differences cause 1e-1 order error (or simply because the model became shallower by removing some batch normalization layers, this model showed smaller error).

But I can't find out any problem in batch nomalization layer implementation of ONNX source, and in any case, 1e-3 error is not acceptable. 

Anyone knows the way to get good result using ONNX model?

## Environment
OS : Ubuntu 18.04
Pytorch version : 1.3.0
Cuda version : 9.0
CuDNN version : 7.6.5
ONNX version : 1.6.0
CPU : i5 9400F
GPU : 1080Ti

Sorry for bad English.
Thanks.

cc @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

**Comment by xsacha** [comment]
> Convert from BGR to RGB (for ONNX model only)

Why for ONNX only?

**Comment by colesbury** [comment]
It sounds like something might be wrong with your preprocessing. Check the outputs of the first layer. If those match, keep checking outputs of subsequent layers.

I'm not sure there's enough information here for other people to help debug your ONNX conversion.

**Comment by fsi-tk** [comment]
> > Convert from BGR to RGB (for ONNX model only)
> 
> Why for ONNX only?
> 
> I guess it is done here in PyTorch version:
> https://github.com/KaiyangZhou/deep-person-reid/blob/6b8fe56638d81c36b7872e8e41efe18232335f2f/torchreid/utils/tools.py#L111

Thank you for reply!

You are right.
Even though I surely converted BGR image to RGB for both models, I overlooked the code you point to me (so I edited my question. For both models, image is converted).
On that assumption, anyone know good way to minimize error?

**Comment by fsi-tk** [comment]
@colesbury

> It sounds like something might be wrong with your preprocessing. Check the outputs of the first layer. If those match, keep checking outputs of subsequent layers.
> 
> I'm not sure there's enough information here for other people to help debug your ONNX conversion.

Thank you for reply!

This is my ONNX conversion code. I use this script the directory where I clone https://github.com/KaiyangZhou/deep-person-reid.

```
# pth2onnx.sh

python pth2onnx.py\
    model.name osnet_x1_0\                                                                              
    model.load_weights .osnet/model.pth.tar-60
```

```
# pth2onnx.py

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import sys
import os
import os.path as osp
import warnings
import time
import argparse

import json
import cv2
import numpy as np                                                                          
import time
import datetime

import os
import torch
import torch.nn as nn
import onnx

from scripts.default_config import (
    get_default_config, imagedata_kwargs, videodata_kwargs,
)
import torchreid
from torchreid.utils import (
    Logger, set_random_seed, check_isfile, resume_from_checkpoint,
    load_pretrained_weights, compute_model_complexity, collect_env_info
)

def build_datamanager(cfg):
    if cfg.data.type == 'image':
        return torchreid.data.ImageDataManager(**imagedata_kwargs(cfg))
    else:
        return torchreid.data.VideoDataManager(**videodata_kwargs(cfg))

def build_engine(cfg, datamanager, model, optimizer, scheduler):
    if cfg.data.type == 'image':
        if cfg.loss.name == 'softmax':
            engine = torchreid.engine.ImageSoftmaxEngine(
                datamanager,
                model,
                optimizer,
                scheduler=scheduler,
                use_gpu=cfg.use_gpu,
                label_smooth=cfg.loss.softmax.label_smooth
            )
        else:
            engine = torchreid.engine.ImageTripletEngine(
                datamanager,
                model,
                optimizer,
                margin=cfg.loss.triplet.margin,
                weight_t=cfg.loss.triplet.weight_t,
                weight_x=cfg.loss.triplet.weight_x,
                scheduler=scheduler,
                use_gpu=cfg.use_gpu,
                label_smooth=cfg.loss.softmax.label_smooth
            )
    
    else:
        if cfg.loss.name == 'softmax':
            engine = torchreid.engine.VideoSoftmaxEngine(
                datamanager,
                model,
                optimizer,
                scheduler=scheduler,
                use_gpu=cfg.use_gpu,
                label_smooth=cfg.loss.softmax.label_smooth,
                pooling_method=cfg.video.pooling_method
            )
        else:
            engine = torchreid.engine.VideoTripletEngine(
                datamanager,
                model,
                optimizer,
                margin=cfg.loss.triplet.margin,
                weight_t=cfg.loss.triplet.weight_t,
                weight_x=cfg.loss.triplet.weight_x,
                scheduler=scheduler,
                use_gpu=cfg.use_gpu,
                label_smooth=cfg.loss.softmax.label_smooth
            )

    return engine

def reset_config(cfg, args):
    if args.root:
        cfg.data.root = args.root
    if args.sources:
        cfg.data.sources = args.sources
    if args.targets:
        cfg.data.targets = args.targets
    if args.transforms:
        cfg.data.transforms = args.transforms

def main():
    input_h = 256
    input_w = 128

    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('--config-file', type=str, default='', help='path to config file')
    parser.add_argument('-s', '--sources', type=str, nargs='+', help='source datasets (delimited by space)')
    parser.add_argument('-t', '--targets', type=str, nargs='+', help='target datasets (delimited by space)')
    parser.add_argument('--transforms', type=str, nargs='+', help='data augmentation')
    parser.add_argument('--root', type=str, default='', help='path to data root')
    parser.add_argument('--gpu-devices', type=str, default='',)
    parser.add_argument('opts', default=None, nargs=argparse.REMAINDER, help='Modify config options using the command-line')
    args = parser.parse_args()

    cfg = get_default_config()
    cfg.use_gpu = torch.cuda.is_available()
    if args.config_file:
        cfg.merge_from_file(args.config_file)
    reset_config(cfg, args)
    cfg.merge_from_list(args.opts)
    set_random_seed(cfg.train.seed)

    if cfg.use_gpu and args.gpu_devices:
        # if gpu_devices is not specified, all available gpus will be used
        os.environ['CUDA_VISIBLE_DEVICES'] = args.gpu_devices
    log_name = 'test.log' if cfg.test.evaluate else 'train.log'
    log_name += time.strftime('-%Y-%m-%d-%H-%M-%S')
    sys.stdout = Logger(osp.join(cfg.data.save_dir, log_name))
    
    print('Show configuration\n{}\n'.format(cfg))
    print('Collecting env info ...')
    print('** System info **\n{}\n'.format(collect_env_info()))
    
    if cfg.use_gpu:
        torch.backends.cudnn.benchmark = True
    
    datamanager = build_datamanager(cfg)
    
    print('Building model: {}'.format(cfg.model.name))
    model = torchreid.models.build_model(
        name=cfg.model.name,
        num_classes=datamanager.num_train_pids,
        loss=cfg.loss.name,
        pretrained=cfg.model.pretrained,
        use_gpu=cfg.use_gpu
    )

    num_params, flops = compute_model_complexity(model, (1, 3, cfg.data.height, cfg.data.width))
    print('Model complexity: params={:,} flops={:,}'.format(num_params, flops))

    if cfg.model.load_weights and check_isfile(cfg.model.load_weights):
        load_pretrained_weights(model, cfg.model.load_weights)

    model = model.cuda()

    print("Start converting !")

    os.environ['CUDA_VISIBLE_DEVICES'] = "0"

    test_size = 1
    input_size = 1 * 3 * input_h * input_w
    print('input_h : ', input_h)
    print('input_w : ', input_w)
    X_test = np.ones((test_size, input_size), dtype=np.float32)

    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    print('Device used:', device)
    dummy_input = torch.from_numpy(X_test[0].reshape(1, 3, input_h, input_w)).float().to(device)

    torch.onnx.export(model, dummy_input, './model/osnet.onnx', input_names=['input'], output_names=['output'], verbose=True, export_params=True, opset_version=9)
```

If necessary, I add more information.

**Comment by fsi-tk** [comment]
And my ONNX execution code which contains pre-processing is here.

```
import onnxruntime
import cv2
import numpy as np

img_path = "./reid-data/market1501/Market-1501-v15.09.15/query/0240_c3s1_049976_00.jpg"
mean = np.array([0.485, 0.456, 0.406])
std  = np.array([0.229, 0.224, 0.225])
image = cv2.imread(img_path, cv2.IMREAD_COLOR)
image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
image = cv2.resize(image, (128, 256)).astype(np.float32)
print(image[0][0][0])
image = ((image / 255.0) - mean ) / std
print("image.shape")
print(image.shape)
img_data = np.expand_dims(image, 0)
img_data = np.transpose(img_data, [0, 3, 1, 2]) 
print("image.shape")
print(img_data.shape)

model_path = "./model/osnet.onnx"
session_option = onnxruntime.SessionOptions()
session_option.log_severity_level = 4
model = onnxruntime.InferenceSession(model_path, sess_options=session_option)
ort_inputs_name = model.get_inputs()[0].name
ort_ouputs_names = [out.name for out in model.get_outputs()]

ort_outs = model.run(ort_ouputs_names, {ort_inputs_name: img_data.astype('float32')})
outputs = np.array(ort_outs[0]).astype("float32")
print(outputs)
```

**Comment by fsi-tk** [comment]
I have solved this problem, using OpenCV instead of Pillow for original Pytorch model.
(The differences between OpenCV and Pillow might affect output's error)
So, I close this issue.

**Comment by Hwijune** [comment]
> 원래 Pytorch 모델에 대해 Pillow 대신 OpenCV를 사용하여이 문제를 해결했습니다.
> (OpenCV와 Pillow의 차이점은 출력 오류에 영향을 줄 수 있습니다)
> 따라서이 문제를 닫습니다.

hi @fsi-tk 

I want to convert model onet_ain_x1_0 to onnx.

osnet_ain_x1_0
osnet_ain_x1_0_msmt17_256x128_amsgrad_ep50_lr0.0015_coslr_b64_fb10_softmax_labsmth_flip_jitter.pth

how can i use your source code??

**Comment by HefnySco** [comment]
> original



> I have solved this problem, using OpenCV instead of Pillow for original Pytorch model. (The differences between OpenCV and Pillow might affect output's error) So, I close this issue.

please explain. I face the same issue when exporting yolo5 pytorch model to onnex format.
